File: src/catalog/freshness.ts
/**
 * Catalog freshness: how old the committed catalog is, and whether a refresh
 * against the offline registry snapshot would change it. Pure and deterministic
 * — it reads the catalog's `generatedAt` and an enrich diff, and makes no I/O.
 * The weekly freshness pipeline uses this to alert maintainers without ever
 * mutating `data/models.json` or making a network call from an advice path.
 */
import type { EnrichDiff } from "./enrich.js";
import type { Catalog } from "../types.js";

/** Days after which the committed catalog is considered stale and should be refreshed. */
export const CATALOG_STALE_AFTER_DAYS = 7;

const MS_PER_DAY = 24 * 60 * 60 * 1000;

export type FreshnessStatus = "fresh" | "stale";

/**
 * Whole-day age of a catalog `generatedAt` timestamp relative to `now`. Throws on
 * an unparseable timestamp (honesty gate — never guess an age) and clamps a
 * future timestamp to 0 so clock skew can't read as "fresh" by accident.
 */
export function catalogAgeDays(generatedAt: string, now: Date): number {
  const generatedMs = new Date(generatedAt).getTime();
  if (Number.isNaN(generatedMs)) {
    throw new Error(`catalog generatedAt is not a valid date: ${generatedAt}`);
  }
  return Math.max(0, Math.floor((now.getTime() - generatedMs) / MS_PER_DAY));
}

/** Counts of what a refresh against the committed snapshot would change. */
export interface DriftCounts {
  readonly added: number;
  readonly updated: number;
  readonly removed: number;
  readonly skipped: number;
  readonly capped: number;
}

export interface FreshnessReport {
  readonly generatedAt: string;
  readonly ageDays: number;
  readonly staleAfterDays: number;
  readonly status: FreshnessStatus;
  readonly hasDrift: boolean;
  readonly drift: DriftCounts;
  /** True when a maintainer should refresh: stale, or actionable drift exists. */
  readonly needsAttention: boolean;
  /** Human-readable reasons; empty when fresh and in sync. */
  readonly reasons: readonly string[];
}

export interface FreshnessInput {
  readonly catalog: Catalog;
  readonly diff: EnrichDiff;
  readonly now: Date;
  readonly staleAfterDays?: number;
}

/** Evaluate catalog freshness + drift into a single actionable report. */
export function evaluateCatalogFreshness(input: FreshnessInput): FreshnessReport {
  const staleAfterDays = input.staleAfterDays ?? CATALOG_STALE_AFTER_DAYS;
  const ageDays = catalogAgeDays(input.catalog.generatedAt, input.now);
  const drift: DriftCounts = {
    added: input.diff.added.length,
    updated: input.diff.updated.length,
    removed: input.diff.removed.length,
    skipped: input.diff.skipped.length,
    capped: input.diff.capped.length,
  };
  // Actionable drift = real catalog changes only. License-gated skips and
  // size-cap drops are intentional and not something a refresh should chase.
  const hasDrift = drift.added > 0 || drift.updated > 0 || drift.removed > 0;
  const status: FreshnessStatus = ageDays > staleAfterDays ? "stale" : "fresh";

  const reasons: string[] = [];
  if (status === "stale") {
    reasons.push(`catalog is ${String(ageDays)} days old (stale after ${String(staleAfterDays)})`);
  }
  if (hasDrift) {
    reasons.push(
      `registry snapshot yields drift: +${String(drift.added)} ~${String(
        drift.updated,
      )} -${String(drift.removed)}`,
    );
  }

  return {
    generatedAt: input.catalog.generatedAt,
    ageDays,
    staleAfterDays,
    status,
    hasDrift,
    drift,
    needsAttention: status === "stale" || hasDrift,
    reasons,
  };
}

/** Render a freshness report as human-readable lines (for logs and issue bodies). */
export function formatFreshnessReport(report: FreshnessReport): string {
  const lines = [
    "Catalog freshness",
    `  generated: ${report.generatedAt}`,
    `  age:       ${String(report.ageDays)} day(s) (stale after ${String(report.staleAfterDays)})`,
    `  status:    ${report.status}`,
    `  drift:     added=${String(report.drift.added)} updated=${String(
      report.drift.updated,
    )} removed=${String(report.drift.removed)} skipped=${String(
      report.drift.skipped,
    )} capped=${String(report.drift.capped)}`,
    `  attention: ${report.needsAttention ? "yes" : "no"}`,
  ];
  if (report.reasons.length > 0) {
    lines.push("  reasons:");
    for (const reason of report.reasons) {
      lines.push(`    - ${reason}`);
    }
  }
  return lines.join("\n");
}

File: crates/llmup-core/tests/freshness.rs
use llmup_core::enrich::EnrichDiff;
use llmup_core::freshness::{evaluate, format_report};

const GENERATED: &str = "2026-01-01T00:00:00.000Z";

#[test]
fn each_actionable_change_independently_requires_attention() {
    for diff in [
        EnrichDiff {
            added: vec!["new:8b".into()],
            ..EnrichDiff::default()
        },
        EnrichDiff {
            updated: vec!["updated:8b".into()],
            ..EnrichDiff::default()
        },
        EnrichDiff {
            removed: vec!["removed:8b".into()],
            ..EnrichDiff::default()
        },
    ] {
        let report = evaluate(GENERATED, &diff, GENERATED, 7).unwrap();
        let json = serde_json::to_value(report).unwrap();
        assert_eq!(json["status"], "fresh");
        assert_eq!(json["hasDrift"], true);
        assert_eq!(json["needsAttention"], true);
        assert_eq!(json["reasons"].as_array().unwrap().len(), 1);
    }
}

#[test]
fn whole_day_age_and_threshold_match_retained_contract() {
    for (now, age, status) in [
        ("2025-12-01T00:00:00Z", 0, "fresh"),
        ("2026-01-02T23:59:59Z", 1, "fresh"),
        ("2026-01-08T23:59:59Z", 7, "fresh"),
        ("2026-01-09T00:00:00Z", 8, "stale"),
    ] {
        let report = evaluate(GENERATED, &EnrichDiff::default(), now, 7).unwrap();
        let json = serde_json::to_value(&report).unwrap();
        assert_eq!(json["ageDays"], age);
        assert_eq!(json["status"], status);
        assert_eq!(json["needsAttention"], status == "stale");
    }
    assert!(evaluate("invalid", &EnrichDiff::default(), GENERATED, 7).is_err());
    assert!(evaluate(GENERATED, &EnrichDiff::default(), "invalid", 7).is_err());
    let custom = evaluate(GENERATED, &EnrichDiff::default(), "2026-01-08T00:00:00Z", 5).unwrap();
    assert_eq!(serde_json::to_value(custom).unwrap()["status"], "stale");
}

#[test]
fn skipped_and_capped_models_are_not_actionable() {
    let diff = EnrichDiff {
        skipped: vec!["closed:70b".into()],
        capped: vec!["tiny:1b".into()],
        ..EnrichDiff::default()
    };
    let report = evaluate(GENERATED, &diff, "2026-01-05T00:00:00Z", 7).unwrap();
    assert_eq!(
        serde_json::to_value(&report).unwrap(),
        serde_json::json!({
            "generatedAt": GENERATED, "ageDays": 4, "staleAfterDays": 7,
            "status": "fresh", "hasDrift": false, "needsAttention": false,
            "drift": {"added": 0, "updated": 0, "removed": 0, "skipped": 1, "capped": 1},
            "reasons": []
        })
    );
    assert!(!format_report(&report).contains("reasons:"));
}

#[test]
fn exact_json_and_human_report_preserve_both_attention_reasons() {
    let diff = EnrichDiff {
        added: vec!["qwen3:32b".into()],
        updated: vec!["llama3.1:8b".into()],
        removed: vec!["withdrawn:8b".into()],
        ..EnrichDiff::default()
    };
    let report = evaluate(GENERATED, &diff, "2026-03-01T00:00:00Z", 7).unwrap();
    assert_eq!(
        serde_json::to_value(&report).unwrap(),
        serde_json::json!({
            "generatedAt": GENERATED, "ageDays": 59, "staleAfterDays": 7,
            "status": "stale", "hasDrift": true, "needsAttention": true,
            "drift": {"added": 1, "updated": 1, "removed": 1, "skipped": 0, "capped": 0},
            "reasons": ["catalog is 59 days old (stale after 7)", "registry snapshot yields drift: +1 ~1 -1"]
        })
    );
    assert_eq!(
        format_report(&report),
        "Catalog freshness\n  generated: 2026-01-01T00:00:00.000Z\n  age:       59 day(s) (stale after 7)\n  status:    stale\n  drift:     added=1 updated=1 removed=1 skipped=0 capped=0\n  attention: yes\n  reasons:\n    - catalog is 59 days old (stale after 7)\n    - registry snapshot yields drift: +1 ~1 -1"
    );
    let fresh = evaluate(GENERATED, &diff, GENERATED, 7).unwrap();
    assert_eq!(serde_json::to_value(fresh).unwrap()["needsAttention"], true);
}
